### PR TITLE
Replace __APPLE__ with __clang__ where applicable

### PR DESCRIPTION
**Pull request by JunielKatarn**
### Type of Change
- Bug fix (non-breaking change which fixes an issue)

### Why
Some compilation scenarios fail when using the Clang/LLVM compiler collection on Windows.

Resolves (internal)

### What
Check for Clang instead of Apple for some compilation compatibility scenarios.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/react-native-windows/pull/9461)

### CODE REVIEW

File: vnext/Chakra/ChakraValue.cpp
Comment (ZacharyHenkel) [inline vnext/Chakra/ChakraValue.cpp:12]: This may be correct as previously authored.

File: vnext/Microsoft.ReactNative.Cxx/NativeModules.h
Comment (ZacharyHenkel) [inline vnext/Microsoft.ReactNative.Cxx/NativeModules.h:304]: This is the location that's causing issues

File: vnext/Microsoft.ReactNative/JsiReader.cpp
Comment (ZacharyHenkel) [inline vnext/Microsoft.ReactNative/JsiReader.cpp:9]: This, and the other instances that refer to the "crash.h" header probably more correct as `__APPLE__` since the header is either an OS header or specific to the Apple standard library

File: vnext/Microsoft.ReactNative/JsiReader.h
Comment (ZacharyHenkel) [inline vnext/Microsoft.ReactNative/JsiReader.h:16]: This looks like a standard library issue and should be reverted

**Review by ZacharyHenkel — CHANGES_REQUESTED** [review]

**Comment by ZacharyHenkel** [comment]
LGTM (even though I'm not a Code Owner)

**Comment by JunielKatarn** [comment]
@ZihanChen-MSFT the change affects code you introduced in #5974.

Can you please review and confirm this is adequate (testing for Clang instead of Apple)?

**Review by NickGerleman — APPROVED** [review]

**Review by NikoAri — APPROVED** [review]
:shipit:

**Comment by NickGerleman** [comment]
Talked offline about backporting. This change helps to unblock an Office-wide static analysis project, for quite a few downstream packages.

Office is validating the move to 0.66 for all platforms, but I'm not sure of the timeline. @HeyImChris, @acoates-ms, do you have a sense of when that transition might happen. @ZacharyHenkel, do you have a sense of when this would be useful to have by?

File: vnext/Microsoft.ReactNative.Cxx/NativeModules.h
Comment (vmoroz) [inline vnext/Microsoft.ReactNative.Cxx/NativeModules.h:305]: >__FUNCSIG__ [](http://example.com/codeflow?start=13&length=11)

This is MSVC specific macro. I am guilty of adding it and not properly #ifdef-ing it. 
Though instead of using __clang__ we must use a test for MSVC compiler because this code would fail in any other compiler except MSVC.

**Review by vmoroz — APPROVED** [review]
:shipit:

File: vnext/Microsoft.ReactNative.Cxx/NativeModules.h
Comment (ZacharyHenkel) [reply vnext/Microsoft.ReactNative.Cxx/NativeModules.h:305]: No because _MSC_VER is also defined when compiling with clang-cl.  Internally we've used a combination of __clang__ and MS_TARGET_WINDOWS which doesn't help in this case.  I'd leave it as is until we prove we have consumers of other compilers.

Comment (JunielKatarn) [reply vnext/Microsoft.ReactNative.Cxx/NativeModules.h:305]: Merging as is, then.

**Pull request merged**
